<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Interactive visualization of scRNA-seq data with Cerebro • cerebroApp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Interactive visualization of scRNA-seq data with Cerebro">
<meta property="og:description" content="Cerebro is a Shiny application that allows to interactively visualize scRNA-seq data. Data must be exported from a Seurat object using the helper functions which also allows to perform analysis such as pathway enrichment analysis based on marker genes of samples and/or clusters.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">cerebroApp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/cerebroApp_workflow.html">The cerebroApp workflow</a>
    </li>
    <li>
      <a href="articles/Cerebro_interface.html">The Cerebro interface</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/romanhaa/cerebroApp">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">


<p><img align="right" width="35%" height="auto" src="articles/logo_Cerebro.png"></p>
<div id="cerebroapp" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#cerebroapp" class="anchor"></a>cerebroApp</h1></div>
<p>R package upon which the <a href="https://github.com/romanhaa/Cerebro">Cerebro</a> is built. Contains helper function that prepare single-cell RNA-seq data stored in a Seurat object for visualization in Cerebro. Both Seurat v2 and Seurat v3 are supported.</p>
<p>Make sure to install the package using <code><a href="https://rdrr.io/pkg/BiocManager/man/install.html">BiocManager::install()</a></code> because <code><a href="https://rdrr.io/pkg/devtools/man/remote-reexports.html">devtools::install_github()</a></code> will otherwise pull old versions of dependencies that can result in errors.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>BiocManager<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">'romanhaa/cerebroApp'</span>)</span></code></pre></div>
<p>For further details, please refer to the official <a href="https://github.com/romanhaa/Cerebro">Cerebro</a> repository.</p>
<div id="how-to-use" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-use" class="anchor"></a>How to use</h2>
<div id="export-data-stored-in-seurat-object" class="section level3">
<h3 class="hasAnchor">
<a href="#export-data-stored-in-seurat-object" class="anchor"></a>Export data stored in Seurat object</h3>
<p><strong>Required meta data:</strong></p>
<ul>
<li>Experiment name.</li>
<li>Organism, e.g. ‘hg’ (human) or ‘mm’ (mouse).</li>
<li>Sample.</li>
<li>Cluster.</li>
<li>Number of transcripts (usually created by Seurat by default; <code>nUMI</code> / <code>nCounts_RNA</code> in Seurat v2 and v3).</li>
<li>Number of expressed genes (usually created by Seurat by default; <code>nGene</code> / <code>nFeature_RNA</code> in Seurat v2 and v3).</li>
</ul>
<p><strong>Note:</strong> It is recommended to save sample information in a column called <code>sample</code> and cluster information in a column called <code>cluster</code>. Otherwise, the respective column names need to specified below.</p>
<p>Prepare data:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">'cerebroApp'</span>)</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="kw"><a href="reference/exportFromSeurat.html">exportFromSeurat</a></span>(seurat, <span class="st">'my_experiment.crb'</span>)</span></code></pre></div>
</div>
<div id="launch-cerebro" class="section level3">
<h3 class="hasAnchor">
<a href="#launch-cerebro" class="anchor"></a>Launch Cerebro</h3>
<p>Launch Cerebro and load the <code>.crb</code> (or <code>.rds</code>) file you just exported from R.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="kw"><a href="reference/launchCerebro.html">launchCerebro</a></span>()</span></code></pre></div>
</div>
<div id="optional-but-recommended-steps" class="section level3">
<h3 class="hasAnchor">
<a href="#optional-but-recommended-steps" class="anchor"></a>Optional (but recommended) steps</h3>
<p>To take full advantage of Cerebro, it is recommended to also run the commands below before exporting the data.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>seurat &lt;-<span class="st"> </span><span class="kw"><a href="reference/addPercentMtRibo.html">addPercentMtRibo</a></span>(seurat)</span>
<span id="cb4-2"><a href="#cb4-2"></a>seurat &lt;-<span class="st"> </span><span class="kw"><a href="reference/getMostExpressedGenes.html">getMostExpressedGenes</a></span>(seurat)</span>
<span id="cb4-3"><a href="#cb4-3"></a>seurat &lt;-<span class="st"> </span><span class="kw"><a href="reference/getMarkerGenes.html">getMarkerGenes</a></span>(seurat)</span>
<span id="cb4-4"><a href="#cb4-4"></a>seurat &lt;-<span class="st"> </span><span class="kw"><a href="reference/getEnrichedPathways.html">getEnrichedPathways</a></span>(seurat)</span>
<span id="cb4-5"><a href="#cb4-5"></a>seurat &lt;-<span class="st"> </span><span class="kw"><a href="reference/performGeneSetEnrichmentAnalysis.html">performGeneSetEnrichmentAnalysis</a></span>(seurat, <span class="dt">GMT_file =</span> <span class="st">'c2.all.v7.0.symbols.gmt'</span>)</span></code></pre></div>
</div>
</div>
<div id="credit" class="section level2">
<h2 class="hasAnchor">
<a href="#credit" class="anchor"></a>Credit</h2>
<ul>
<li>Pathway enrichment in marker gene lists (<code><a href="reference/getEnrichedPathways.html">getEnrichedPathways()</a></code>) is done through the enrichR API (<a href="https://github.com/wjawaid/enrichR" class="uri">https://github.com/wjawaid/enrichR</a>). I took the <code>enrichr</code> function and modified it to run in parallel (<code>future_lapply</code>) and not print status messages.</li>
<li>Gene set enrichment analysis (<code><a href="reference/performGeneSetEnrichmentAnalysis.html">performGeneSetEnrichmentAnalysis()</a></code>) is performed using the <a href="https://bioconductor.org/packages/release/bioc/html/GSVA.html">GSVA</a> R package. p- and q-value statistics are calculated through the same method as used by “Evaluation of methods to assign cell type labels to cell clusters from single-cell RNA-sequencing data”, Diaz-Mejia <em>et al</em>., F1000Research (2019). <a href="https://f1000research.com/articles/8-296/v2">Link to publication</a>
</li>
</ul>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/romanhaa/cerebroApp">https://​github.com/​romanhaa/​cerebroApp</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing cerebroApp</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Roman Hillje <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://opensource.org/licenses/MIT"><img src="https://img.shields.io/badge/License-MIT-yellow.svg" alt="License: MIT"></a></li>
<li><a href="https://www.tidyverse.org/lifecycle/#maturing"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle: maturing"></a></li>
<li><a href="https://twitter.com/fakechek1"><img src="https://img.shields.io/twitter/url/https/twitter.com/fakechek1.svg?label=Follow%20%40fakechek1&amp;style=social" alt="Twitter"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Roman Hillje.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
