---
title: Introduction to the cerebroApp workflow
author: Roman Hillje
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    number_sections: true
    self_contained: yes
    mode: selfcontained
vignette: >
  %\VignetteIndexEntry{Introduction to cerebroApp}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

The `cerebroApp` package has two main purposes: (1) Give access to the Cerebro user interface, and (2) provide a set of functions to pre-process and export scRNA-seq data for visualization in Cerebro.
The Cerebro user interface was built using the Shiny framework and designed to provide numerous perspectives on a given data set that should facilitate data interpretation.
This vignette will give you an example of how scRNA-seq data can be processed using the [Seurat](https://satijalab.org/seurat/) toolkit and then exported for visualization in Cerebro.

Ket features that can be accessed through Cerebro:

* Interactive 2D and 3D dimensional reductions.
* Sample and cluster overview panels.
* Tables of most expressed genes and marker genes for samples and clusters.
* Tables of enriched pathways for samples and clusters.
* Query gene(s) and gene sets from MSigDB and show their expression in dimensional reductions.
* Visualize trajectories calculated with Monocle v2.
* All plots can be exported to PNG. In addition, 2D dimensional reductions can be exported to PDF.
* Tables can be downloaded in CSV or Excel format.

# Installation

Before starting the workflow, we need to install cerebroApp, as well as the [Seurat](https://satijalab.org/seurat/) and [monocle](http://cole-trapnell-lab.github.io/monocle-release/) packages, which are not installed as dependencies of cerebroApp because they are only necessary if you want/need to pre-process your scRNA-seq data but not if you just want to launch the Cerebro user interface, e.g. because you already have the pre-processed data.

```{r installation, message=FALSE, warning=FALSE}
if ( 'BiocManager' %in% installed.packages() == FALSE ) install.packages('BiocManager')

library(BiocManager)

if ( 'cerebroApp' %in% installed.packages() == FALSE ) install('cerebroApp')
if ( 'Seurat' %in% installed.packages() == FALSE ) install('Seurat')
if ( 'monocle' %in% installed.packages() == FALSE ) install('monocle')
```

# Example of workflow

## Load packages

Apart from the packages we just installed, we will also load the dplyr and devtools packages and set a seed to make our analysis reproducible.

```{r setup, message=FALSE, warning=FALSE}
library(dplyr)
library(Seurat)
library(monocle)
library(cerebroApp)
library(devtools)

set.seed(1234567)
```

## Load count data

In this example workflow, we will load a small transcript count table from the [Seurat](https://satijalab.org/seurat/) package containing 80 cells and 230 genes.

```{r load_data}
pbmc <- read.table(
  file = system.file('extdata', 'pbmc_raw.txt', package = 'Seurat'),
  as.is = TRUE
)
```

## Pre-processing with Seurat

Now, we create a Seurat object and filter out cells with less than `50` transcripts or fewer than `10` expressed genes.
Those cut-offs are only reasonable for this example data set and will likely need to be adjusted in a real data set.
Then, we follow the [standard Seurat workflow](https://satijalab.org/seurat/essential_commands.html), including...

* log-normalization of transcript counts,
* identification of highly variably genes,
* scaling of expression values and regression for the number of transcripts per cell,
* principal component analysis (PCA).

```{r pre_processing, warning=FALSE}
seurat <- CreateSeuratObject(
  project = 'pbmc',
  counts = pbmc,
  min.cells = 10
)
seurat <- subset(seurat, subset = nCount_RNA > 50 & nFeature_RNA > 10)
seurat <- NormalizeData(seurat, verbose = FALSE)
seurat <- FindVariableFeatures(seurat, verbose = FALSE)
seurat <- ScaleData(seurat, vars.to.regress = 'nCount_RNA', verbose = FALSE)
seurat <- RunPCA(seurat, features = seurat@assays$RNA@var.features, verbose = FALSE)
```

## Clustering

Then, we identify clusters and build a cluster tree that represents the similarity between clusters.

```{r clustering}
seurat <- FindNeighbors(seurat, verbose = FALSE)
seurat <- FindClusters(seurat, resolution = 0.5)
seurat <- BuildClusterTree(
  seurat,
  dims = 1:30,
  reorder = TRUE,
  reorder.numeric = TRUE,
  verbose = FALSE
)
```

## Cell cycle analysis

We could also perform cell cycle analysis using the `CellCycleScoring()` built into Seurat.
However, too few of the cell cycle genes are present in this example data set so we will skip it here.
The code below should work with a real data set.

```{r cell_cycle_analysis, eval=FALSE}
seurat <- CellCycleScoring(
  seurat,
  g2m.features = cc.genes$g2m.genes,
  s.features = cc.genes$s.genes
)

seurat@misc$gene_lists$G2M_phase_genes <- cc.genes$g2m.genes
seurat@misc$gene_lists$S_phase_genes <- cc.genes$s.genes
```

## Dimensional reduction

Next, we use the UMAP technique to generate two- and three-dimensional projections.

```{r dimensional_reduction, warning=FALSE}
seurat <- RunUMAP(
  seurat,
  reduction.name = 'UMAP',
  reduction.key = 'UMAP_',
  dims = 1:30,
  n.components = 2,
  seed.use = 100,
  verbose = FALSE
)

seurat <- RunUMAP(
  seurat,
  reduction.name = 'UMAP_3D',
  reduction.key = 'UMAP3D_',
  dims = 1:30,
  n.components = 3,
  seed.use = 100,
  verbose = FALSE
)
```

## Curate meta data

This example data set consists of a single sample so we just add that name to the meta data.
Moreover, in order to be able to understand how we did the analysis later, we add some meta data to the `misc` slot of our Seurat object.

```{r prepare_meta_data}
seurat@meta.data$sample <- factor('pbmc', levels = 'pbmc')

seurat@misc$experiment <- list(
  experiment_name = 'pbmc',
  organism = 'hg',
  date_of_analysis = Sys.Date()
)

seurat@misc$parameters <- list(
  gene_nomenclature = 'gene_name',
  discard_genes_expressed_in_fewer_cells_than = 10,
  keep_mitochondrial_genes = TRUE,
  variables_to_regress_out = 'nUMI',
  number_PCs = 30,
  tSNE_perplexity = 30,
  cluster_resolution = 0.5
)

seurat@misc$parameters$filtering <- list(
  UMI_min = 50,
  UMI_max = Inf,
  genes_min = 10,
  genes_max = Inf
)

seurat@misc$technical_info <- list(
  'R' = capture.output(session_info())
)
```

## cerebroApp functions (optional but recommended)

Now we use a set of cerebroApp function to get more information about the data set and the contained samples and clusters.

### Calculate percentage of mitochondrial and ribosomal transcripts

First, for every cell we calculate the percentage of mitochondrial and ribosomal transcripts of all transcripts using the `addPercentMtRibo()` function.

```{r addPercentMtRibo}
seurat <- addPercentMtRibo(
  seurat,
  organism = 'hg',
  gene_nomenclature = 'name'
)
```

### Get most expressed genes

Another interesting perspective is to look at the 100 most expressed genes in every sample and cluster.
This can help in understanding the transcriptional activity of a given cell population.

```{r getMostExpressedGenes}
seurat <- getMostExpressedGenes(
  seurat,
  column_sample = 'sample',
  column_cluster = 'seurat_clusters'
)
```

### Identify marker genes

Marker genes are genes which are particularly strongly or weakly expressed in a given cell population, e.g. a sample or cluster.
This information can help to distinguish the role of different cell groups in the data set.
We can identify marker genes using the `getMarkerGenes()` function which internally uses the `Seurat::FindAllMarkers()` function but also stores the results in the Seurat object under the `@misc` slot.

```{r getMarkerGenes}
seurat <- getMarkerGenes(
  seurat,
  organism = 'hg',
  column_sample = 'sample',
  column_cluster = 'seurat_clusters',
  only_pos = TRUE
)
```

### Perform pathway enrichment on marker genes

Using the previously identified marker genes, we can perform a pathway enrichmeny analysis using the `getEnrichedPathways()` function.
The function will internally use the [Enrichr](https://amp.pharm.mssm.edu/Enrichr/) API and store the results in the Seurat object under the `@misc` slot.

```{r getEnrichedPathways}
seurat <- getEnrichedPathways(
  seurat,
  column_sample = 'sample',
  column_cluster = 'seurat_clusters',
  adj_p_cutoff = 0.01,
  max_terms = 100
)
```

### Gene set enrichment

On top of querying the databases of the [Enrichr](https://amp.pharm.mssm.edu/Enrichr/) service, one can also perform gene set enrichment analysis on gene sets provided as a `.gmt` file, e.g. from the [MSigDB](https://www.gsea-msigdb.org/gsea/msigdb/index.jsp).
The `performGeneSetEnrichmentAnalysis()` function uses the [GSVA](https://bioconductor.org/packages/release/bioc/html/GSVA.html) method in combination with additional statistics as published by [Diaz-Mejia et. al.](https://f1000research.com/articles/8-296).

```{r performGeneSetEnrichmentAnalysis, eval=FALSE}
example_gene_set <- system.file("extdata/example_gene_set.gmt", package = "cerebroApp")

seurat <- performGeneSetEnrichmentAnalysis(
  seurat,
  GMT_file = example_gene_set,
  column_sample = 'sample',
  column_cluster = 'seurat_clusters',
  thresh_p_val = 0.05,
  thresh_q_val = 0.1,
  parallel.sz = 1,
  verbose = FALSE
)
```

## Trajectory analysis with Monocle (optional)

Then, we perform trajectory analysis with [Monocle v2](http://cole-trapnell-lab.github.io/monocle-release/) using the previously identified highly variable genes.
We extract the trajectory from the generated Monocle object with the `extractMonocleTrajectory()` function of cerebroApp and attach it to our Seurat object.

```{r analysis_with_monocle}
monocle <- newCellDataSet(
  seurat@assays$RNA@data,
  phenoData = new('AnnotatedDataFrame', data = seurat@meta.data),
  featureData = new('AnnotatedDataFrame', data = data.frame(
    gene_short_name = rownames(seurat@assays$RNA@counts),
    row.names = rownames(seurat@assays$RNA@counts))
  )
)

monocle <- estimateSizeFactors(monocle)
monocle <- estimateDispersions(monocle)
monocle <- setOrderingFilter(monocle, seurat@assays$RNA@var.features)
monocle <- reduceDimension(monocle, max_components = 2, method = 'DDRTree')
monocle <- orderCells(monocle)

seurat <- extractMonocleTrajectory(monocle, seurat, 'highly_variable_genes')
```

## Export to Cerebro format

Finally, we use the `exportFromSeurat()` function of cerebroApp to export our Seurat object to a `.crb` file which can be loaded into Cerebro.

```{r export_data}
exportFromSeurat(
  seurat,
  experiment_name = 'pbmc',
  file = paste0('cerebro_pbmc_', Sys.Date(), '.crb'),
  organism = 'hg',
  column_cluster = 'seurat_clusters',
  column_nUMI = 'nCount_RNA',
  column_nGene = 'nFeature_RNA'
)
```

The Cerebro use interface can be launched using the `launchCerebro()` function.

# Other example data sets

We provide ready-to-load `.crb` files for a set of real data sets processed using either [scanpy](https://icb-scanpy.readthedocs-hosted.com/en/stable/) and [Seurat (v2.x and v3.x)](https://satijalab.org/seurat/) on our GitHub repository: <https://github.com/romanhaa/Cerebro/tree/master/examples>

# Session info

```{r session_info}
sessionInfo()
```
