<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Perform gene set enrichment analysis with GSVA. — performGeneSetEnrichmentAnalysis • cerebroApp</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Perform gene set enrichment analysis with GSVA. — performGeneSetEnrichmentAnalysis" />
<meta property="og:description" content="This function calculates enrichment scores, p- and q-value
statistics for provided gene sets for samples and clusters of cells in given
Seurat object using gene set variation analysis (GSVA). Calculation of p- and
q-values for gene sets is performed as done in &quot;Evaluation of methods to
assign cell type labels to cell clusters from single-cell RNA-sequencing
data&quot;, Diaz-Mejia et al., F1000Research (2019)." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cerebroApp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/introduction_to_cerebroApp.html">Introduction to cerebroApp</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/romanhaa/cerebroApp">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform gene set enrichment analysis with GSVA.</h1>
    <small class="dont-index">Source: <a href='https://github.com/romanhaa/cerebroApp/blob/master/R/performGeneSetEnrichmentAnalysis.R'><code>R/performGeneSetEnrichmentAnalysis.R</code></a></small>
    <div class="hidden name"><code>performGeneSetEnrichmentAnalysis.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function calculates enrichment scores, p- and q-value
statistics for provided gene sets for samples and clusters of cells in given
Seurat object using gene set variation analysis (GSVA). Calculation of p- and
q-values for gene sets is performed as done in "Evaluation of methods to
assign cell type labels to cell clusters from single-cell RNA-sequencing
data", Diaz-Mejia et al., F1000Research (2019).</p>
    </div>

    <pre class="usage"><span class='fu'>performGeneSetEnrichmentAnalysis</span>(
  <span class='no'>object</span>,
  <span class='kw'>assay</span> <span class='kw'>=</span> <span class='st'>"RNA"</span>,
  <span class='no'>GMT_file</span>,
  <span class='kw'>column_sample</span> <span class='kw'>=</span> <span class='st'>"sample"</span>,
  <span class='kw'>column_cluster</span> <span class='kw'>=</span> <span class='st'>"cluster"</span>,
  <span class='kw'>thresh_p_val</span> <span class='kw'>=</span> <span class='fl'>0.05</span>,
  <span class='kw'>thresh_q_val</span> <span class='kw'>=</span> <span class='fl'>0.1</span>,
  <span class='no'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Seurat object.</p></td>
    </tr>
    <tr>
      <th>assay</th>
      <td><p>Assay to pull counts from; defaults to 'RNA'. Only relevant in
Seurat v3.0 or higher since the concept of assays wasn't implemented before.</p></td>
    </tr>
    <tr>
      <th>GMT_file</th>
      <td><p>Path to GMT file containing the gene sets to be tested.
The Broad Institute provides many gene sets which can be downloaded:
http://software.broadinstitute.org/gsea/msigdb/index.jsp</p></td>
    </tr>
    <tr>
      <th>column_sample</th>
      <td><p>Column in object@meta.data that contains information
about sample; defaults to 'sample'.</p></td>
    </tr>
    <tr>
      <th>column_cluster</th>
      <td><p>Column in object@meta.data that contains information
about cluster; defaults to 'cluster'.</p></td>
    </tr>
    <tr>
      <th>thresh_p_val</th>
      <td><p>Threshold for p-value, defaults to 0.05.</p></td>
    </tr>
    <tr>
      <th>thresh_q_val</th>
      <td><p>Threshold for q-value, defaults to 0.1.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Further parameters can be passed to control GSVA::gsva().</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Seurat object with GSVA results for samples and clusters stored in
object@misc$enriched_pathways$GSVA.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Perform gene set enrichment analysis with GSVA.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>pbmc</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata/v1.2/seurat_pbmc.rds"</span>,
  <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"cerebroApp"</span>))
<span class='no'>example_gene_set</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata/example_gene_set.gmt"</span>,
  <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"cerebroApp"</span>)
<span class='no'>pbmc</span> <span class='kw'>&lt;-</span> <span class='fu'>performGeneSetEnrichmentAnalysis</span>(
  <span class='kw'>object</span> <span class='kw'>=</span> <span class='no'>pbmc</span>,
  <span class='kw'>GMT_file</span> <span class='kw'>=</span> <span class='no'>example_gene_set</span>,
  <span class='kw'>column_sample</span> <span class='kw'>=</span> <span class='st'>'sample'</span>,
  <span class='kw'>column_cluster</span> <span class='kw'>=</span> <span class='st'>'seurat_clusters'</span>,
  <span class='kw'>thresh_p_val</span> <span class='kw'>=</span> <span class='fl'>0.05</span>,
  <span class='kw'>thresh_q_val</span> <span class='kw'>=</span> <span class='fl'>0.1</span>,
  <span class='kw'>parallel.sz</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>
)</div><div class='output co'>#&gt; <span class='message'>[15:38:52] Loading gene sets...</span></div><div class='output co'>#&gt; <span class='message'>[15:38:52] Loaded 1 gene sets from GMT file.</span></div><div class='output co'>#&gt; <span class='message'>[15:38:52] Extracting transcript counts...</span></div><div class='output co'>#&gt; <span class='message'>[15:38:52] Removing non-expressed genes...</span></div><div class='output co'>#&gt; <span class='message'>[15:38:52] Only 1 sample in the data set, will skip analysis by sample...</span></div><div class='output co'>#&gt; <span class='message'>[15:38:52] Performing GSVA for clusters...</span></div><div class='output co'>#&gt; <span class='message'>[15:38:52] Only 1 gene set available, therefore no p- and q-values can be calculated and no filtering of the results will be performed.</span></div><div class='output co'>#&gt; <span class='message'>[15:38:52] Only 1 gene set available, therefore no p- and q-values can be calculated and no filtering of the results will be performed.</span></div><div class='output co'>#&gt; <span class='message'>[15:38:52] Only 1 gene set available, therefore no p- and q-values can be calculated and no filtering of the results will be performed.</span></div><div class='output co'>#&gt; <span class='message'>[15:38:52] 3 gene sets passed the thresholds across all clusters.</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Roman Hillje.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


