<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to the cerebroApp workflow • cerebroApp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to the cerebroApp workflow">
<meta property="og:description" content="cerebroApp">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cerebroApp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.2.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cerebroApp_workflow.html">The cerebroApp workflow</a>
    </li>
    <li>
      <a href="../articles/Cerebro_interface.html">The Cerebro interface</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/romanhaa/cerebroApp/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to the cerebroApp workflow</h1>
                        <h4 class="author">Roman Hillje</h4>
            
            <h4 class="date">17 June, 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/romanhaa/cerebroApp/blob/master/vignettes/cerebroApp_workflow.Rmd"><code>vignettes/cerebroApp_workflow.Rmd</code></a></small>
      <div class="hidden name"><code>cerebroApp_workflow.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>The <code>cerebroApp</code> package has two main purposes: (1) Give access to the Cerebro user interface, and (2) provide a set of functions to pre-process and export scRNA-seq data for visualization in Cerebro. The Cerebro user interface was built using the Shiny framework and designed to provide numerous perspectives on a given data set that should facilitate data interpretation. This vignette will give you an example of how scRNA-seq data can be processed using the <a href="https://satijalab.org/seurat/">Seurat</a> toolkit and then exported for visualization in Cerebro.</p>
<p>Ket features that can be accessed through Cerebro:</p>
<ul>
<li>Interactive 2D and 3D dimensional reductions.</li>
<li>Sample and cluster overview panels.</li>
<li>Tables of most expressed genes and marker genes for samples and clusters.</li>
<li>Tables of enriched pathways for samples and clusters.</li>
<li>Query gene(s) and gene sets from MSigDB and show their expression in dimensional reductions.</li>
<li>Visualize trajectories calculated with Monocle v2.</li>
<li>All plots can be exported to PNG. In addition, 2D dimensional reductions can be exported to PDF.</li>
<li>Tables can be downloaded in CSV or Excel format.</li>
</ul>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>Before starting the workflow, we need to install cerebroApp, as well as the <a href="https://satijalab.org/seurat/">Seurat</a> and <a href="http://cole-trapnell-lab.github.io/monocle-release/">monocle</a> packages, which are not installed as dependencies of cerebroApp because they are only necessary if you want/need to pre-process your scRNA-seq data but not if you just want to launch the Cerebro user interface, e.g. because you already have the pre-processed data.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="kw">if</span> ( <span class="st">'BiocManager'</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span>() <span class="kw">==</span> <span class="fl">FALSE</span> ) <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">'BiocManager'</span>)

<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BiocManager</span>)

<span class="kw">if</span> ( <span class="st">'cerebroApp'</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span>() <span class="kw">==</span> <span class="fl">FALSE</span> ) <span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">'cerebroApp'</span>)
<span class="kw">if</span> ( <span class="st">'Seurat'</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span>() <span class="kw">==</span> <span class="fl">FALSE</span> ) <span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">'Seurat'</span>)
<span class="kw">if</span> ( <span class="st">'monocle'</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span>() <span class="kw">==</span> <span class="fl">FALSE</span> ) <span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">'monocle'</span>)</pre></body></html></div>
</div>
<div id="example-of-workflow" class="section level1">
<h1 class="hasAnchor">
<a href="#example-of-workflow" class="anchor"></a>Example of workflow</h1>
<div id="load-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#load-packages" class="anchor"></a>Load packages</h2>
<p>Apart from the packages we just installed, we will also load the dplyr and devtools packages and set a seed to make our analysis reproducible.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dplyr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">Seurat</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">monocle</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">cerebroApp</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">devtools</span>)

<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1234567</span>)</pre></body></html></div>
</div>
<div id="load-count-data" class="section level2">
<h2 class="hasAnchor">
<a href="#load-count-data" class="anchor"></a>Load count data</h2>
<p>In this example workflow, we will load a small transcript count table from the <a href="https://satijalab.org/seurat/">Seurat</a> package containing 80 cells and 230 genes.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">pbmc</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(
  <span class="kw">file</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'pbmc_raw.txt'</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">'Seurat'</span>),
  <span class="kw">as.is</span> <span class="kw">=</span> <span class="fl">TRUE</span>
)</pre></body></html></div>
</div>
<div id="pre-processing-with-seurat" class="section level2">
<h2 class="hasAnchor">
<a href="#pre-processing-with-seurat" class="anchor"></a>Pre-processing with Seurat</h2>
<p>Now, we create a Seurat object and filter out cells with less than <code>50</code> transcripts or fewer than <code>10</code> expressed genes. Those cut-offs are only reasonable for this example data set and will likely need to be adjusted in a real data set. Then, we follow the <a href="https://satijalab.org/seurat/essential_commands.html">standard Seurat workflow</a>, including…</p>
<ul>
<li>log-normalization of transcript counts,</li>
<li>identification of highly variably genes,</li>
<li>scaling of expression values and regression for the number of transcripts per cell,</li>
<li>principal component analysis (PCA).</li>
</ul>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">seurat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/CreateSeuratObject.html">CreateSeuratObject</a></span>(
  <span class="kw">project</span> <span class="kw">=</span> <span class="st">'pbmc'</span>,
  <span class="kw">counts</span> <span class="kw">=</span> <span class="no">pbmc</span>,
  <span class="kw">min.cells</span> <span class="kw">=</span> <span class="fl">10</span>
)
<span class="no">seurat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/subset.Seurat.html">subset</a></span>(<span class="no">seurat</span>, <span class="kw">subset</span> <span class="kw">=</span> <span class="no">nCount_RNA</span> <span class="kw">&gt;</span> <span class="fl">50</span> <span class="kw">&amp;</span> <span class="no">nFeature_RNA</span> <span class="kw">&gt;</span> <span class="fl">10</span>)
<span class="no">seurat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/NormalizeData.html">NormalizeData</a></span>(<span class="no">seurat</span>, <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="no">seurat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindVariableFeatures.html">FindVariableFeatures</a></span>(<span class="no">seurat</span>, <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="no">seurat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/ScaleData.html">ScaleData</a></span>(<span class="no">seurat</span>, <span class="kw">vars.to.regress</span> <span class="kw">=</span> <span class="st">'nCount_RNA'</span>, <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="no">seurat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunPCA.html">RunPCA</a></span>(<span class="no">seurat</span>, <span class="kw">features</span> <span class="kw">=</span> <span class="no">seurat</span>@<span class="kw">assays</span>$<span class="no">RNA</span>@<span class="kw">var.features</span>, <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
</div>
<div id="clustering" class="section level2">
<h2 class="hasAnchor">
<a href="#clustering" class="anchor"></a>Clustering</h2>
<p>Then, we identify clusters and build a cluster tree that represents the similarity between clusters.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">seurat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindNeighbors.html">FindNeighbors</a></span>(<span class="no">seurat</span>, <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="no">seurat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindClusters.html">FindClusters</a></span>(<span class="no">seurat</span>, <span class="kw">resolution</span> <span class="kw">=</span> <span class="fl">0.5</span>)
<span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Number of nodes: 77</span>
<span class="co">#&gt; Number of edges: 2280</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Running Louvain algorithm...</span>
<span class="co">#&gt; Maximum modularity in 10 random starts: 0.5740</span>
<span class="co">#&gt; Number of communities: 2</span>
<span class="co">#&gt; Elapsed time: 0 seconds</span>
<span class="no">seurat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/BuildClusterTree.html">BuildClusterTree</a></span>(
  <span class="no">seurat</span>,
  <span class="kw">dims</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fl">30</span>,
  <span class="kw">reorder</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
  <span class="kw">reorder.numeric</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
  <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>
)</pre></body></html></div>
</div>
<div id="cell-cycle-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-cycle-analysis" class="anchor"></a>Cell cycle analysis</h2>
<p>We could also perform cell cycle analysis using the <code><a href="https://rdrr.io/pkg/Seurat/man/CellCycleScoring.html">CellCycleScoring()</a></code> built into Seurat. However, too few of the cell cycle genes are present in this example data set so we will skip it here. The code below should work with a real data set.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">seurat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/CellCycleScoring.html">CellCycleScoring</a></span>(
  <span class="no">seurat</span>,
  <span class="kw">g2m.features</span> <span class="kw">=</span> <span class="no">cc.genes</span>$<span class="no">g2m.genes</span>,
  <span class="kw">s.features</span> <span class="kw">=</span> <span class="no">cc.genes</span>$<span class="no">s.genes</span>
)

<span class="no">seurat</span>@<span class="kw">misc</span>$<span class="no">gene_lists</span>$<span class="no">G2M_phase_genes</span> <span class="kw">&lt;-</span> <span class="no">cc.genes</span>$<span class="no">g2m.genes</span>
<span class="no">seurat</span>@<span class="kw">misc</span>$<span class="no">gene_lists</span>$<span class="no">S_phase_genes</span> <span class="kw">&lt;-</span> <span class="no">cc.genes</span>$<span class="no">s.genes</span></pre></body></html></div>
</div>
<div id="dimensional-reduction" class="section level2">
<h2 class="hasAnchor">
<a href="#dimensional-reduction" class="anchor"></a>Dimensional reduction</h2>
<p>Next, we use the UMAP technique to generate two- and three-dimensional projections.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">seurat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunUMAP.html">RunUMAP</a></span>(
  <span class="no">seurat</span>,
  <span class="kw">reduction.name</span> <span class="kw">=</span> <span class="st">'UMAP'</span>,
  <span class="kw">reduction.key</span> <span class="kw">=</span> <span class="st">'UMAP_'</span>,
  <span class="kw">dims</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fl">30</span>,
  <span class="kw">n.components</span> <span class="kw">=</span> <span class="fl">2</span>,
  <span class="kw">seed.use</span> <span class="kw">=</span> <span class="fl">100</span>,
  <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>
)

<span class="no">seurat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunUMAP.html">RunUMAP</a></span>(
  <span class="no">seurat</span>,
  <span class="kw">reduction.name</span> <span class="kw">=</span> <span class="st">'UMAP_3D'</span>,
  <span class="kw">reduction.key</span> <span class="kw">=</span> <span class="st">'UMAP3D_'</span>,
  <span class="kw">dims</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fl">30</span>,
  <span class="kw">n.components</span> <span class="kw">=</span> <span class="fl">3</span>,
  <span class="kw">seed.use</span> <span class="kw">=</span> <span class="fl">100</span>,
  <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>
)</pre></body></html></div>
</div>
<div id="curate-meta-data" class="section level2">
<h2 class="hasAnchor">
<a href="#curate-meta-data" class="anchor"></a>Curate meta data</h2>
<p>This example data set consists of a single sample so we just add that name to the meta data. Moreover, in order to be able to understand how we did the analysis later, we add some meta data to the <code>misc</code> slot of our Seurat object.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">seurat</span>@<span class="kw">meta.data</span>$<span class="no">sample</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="st">'pbmc'</span>, <span class="kw">levels</span> <span class="kw">=</span> <span class="st">'pbmc'</span>)

<span class="no">seurat</span>@<span class="kw">misc</span>$<span class="no">experiment</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(
  <span class="kw">experiment_name</span> <span class="kw">=</span> <span class="st">'pbmc'</span>,
  <span class="kw">organism</span> <span class="kw">=</span> <span class="st">'hg'</span>,
  <span class="kw">date_of_analysis</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html">Sys.Date</a></span>()
)

<span class="no">seurat</span>@<span class="kw">misc</span>$<span class="no">parameters</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(
  <span class="kw">gene_nomenclature</span> <span class="kw">=</span> <span class="st">'gene_name'</span>,
  <span class="kw">discard_genes_expressed_in_fewer_cells_than</span> <span class="kw">=</span> <span class="fl">10</span>,
  <span class="kw">keep_mitochondrial_genes</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
  <span class="kw">variables_to_regress_out</span> <span class="kw">=</span> <span class="st">'nUMI'</span>,
  <span class="kw">number_PCs</span> <span class="kw">=</span> <span class="fl">30</span>,
  <span class="kw">tSNE_perplexity</span> <span class="kw">=</span> <span class="fl">30</span>,
  <span class="kw">cluster_resolution</span> <span class="kw">=</span> <span class="fl">0.5</span>
)

<span class="no">seurat</span>@<span class="kw">misc</span>$<span class="no">parameters</span>$<span class="no">filtering</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(
  <span class="kw">UMI_min</span> <span class="kw">=</span> <span class="fl">50</span>,
  <span class="kw">UMI_max</span> <span class="kw">=</span> <span class="fl">Inf</span>,
  <span class="kw">genes_min</span> <span class="kw">=</span> <span class="fl">10</span>,
  <span class="kw">genes_max</span> <span class="kw">=</span> <span class="fl">Inf</span>
)

<span class="no">seurat</span>@<span class="kw">misc</span>$<span class="no">technical_info</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(
  <span class="st">'R'</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/capture.output.html">capture.output</a></span>(<span class="fu"><a href="https://devtools.r-lib.org//reference/reexports.html">session_info</a></span>())
)</pre></body></html></div>
</div>
<div id="cerebroapp-functions-optional-but-recommended" class="section level2">
<h2 class="hasAnchor">
<a href="#cerebroapp-functions-optional-but-recommended" class="anchor"></a>cerebroApp functions (optional but recommended)</h2>
<p>Now we use a set of cerebroApp function to get more information about the data set and the contained samples and clusters.</p>
<div id="calculate-percentage-of-mitochondrial-and-ribosomal-transcripts" class="section level3">
<h3 class="hasAnchor">
<a href="#calculate-percentage-of-mitochondrial-and-ribosomal-transcripts" class="anchor"></a>Calculate percentage of mitochondrial and ribosomal transcripts</h3>
<p>First, for every cell we calculate the percentage of mitochondrial and ribosomal transcripts of all transcripts using the <code><a href="../reference/addPercentMtRibo.html">addPercentMtRibo()</a></code> function.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">seurat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/addPercentMtRibo.html">addPercentMtRibo</a></span>(
  <span class="no">seurat</span>,
  <span class="kw">organism</span> <span class="kw">=</span> <span class="st">'hg'</span>,
  <span class="kw">gene_nomenclature</span> <span class="kw">=</span> <span class="st">'name'</span>
)
<span class="co">#&gt; [16:13:15] No mitochondrial genes found in data set.</span>
<span class="co">#&gt; [16:13:15] Calculate percentage of 1 ribosomal transcript(s) present in the data set...</span></pre></body></html></div>
</div>
<div id="get-most-expressed-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#get-most-expressed-genes" class="anchor"></a>Get most expressed genes</h3>
<p>Another interesting perspective is to look at the 100 most expressed genes in every sample and cluster. This can help in understanding the transcriptional activity of a given cell population.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">seurat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/getMostExpressedGenes.html">getMostExpressedGenes</a></span>(
  <span class="no">seurat</span>,
  <span class="kw">column_sample</span> <span class="kw">=</span> <span class="st">'sample'</span>,
  <span class="kw">column_cluster</span> <span class="kw">=</span> <span class="st">'seurat_clusters'</span>
)
<span class="co">#&gt; [16:13:15] Get most expressed genes for 1 sample(s)...</span>
<span class="co">#&gt; [16:13:15] Get most expressed genes for 2 clusters...</span></pre></body></html></div>
</div>
<div id="identify-marker-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#identify-marker-genes" class="anchor"></a>Identify marker genes</h3>
<p>Marker genes are genes which are particularly strongly or weakly expressed in a given cell population, e.g. a sample or cluster. This information can help to distinguish the role of different cell groups in the data set. We can identify marker genes using the <code><a href="../reference/getMarkerGenes.html">getMarkerGenes()</a></code> function which internally uses the <code><a href="https://rdrr.io/pkg/Seurat/man/FindAllMarkers.html">Seurat::FindAllMarkers()</a></code> function but also stores the results in the Seurat object under the <code>@misc</code> slot.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">seurat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/getMarkerGenes.html">getMarkerGenes</a></span>(
  <span class="no">seurat</span>,
  <span class="kw">organism</span> <span class="kw">=</span> <span class="st">'hg'</span>,
  <span class="kw">column_sample</span> <span class="kw">=</span> <span class="st">'sample'</span>,
  <span class="kw">column_cluster</span> <span class="kw">=</span> <span class="st">'seurat_clusters'</span>,
  <span class="kw">only_pos</span> <span class="kw">=</span> <span class="fl">TRUE</span>
)
<span class="co">#&gt; Cache found</span>
<span class="co">#&gt; [16:13:19] Sample column provided but only 1 sample found.</span>
<span class="co">#&gt; [16:13:19] Get marker genes for 2 clusters...</span>
<span class="co">#&gt; Calculating cluster 0</span>
<span class="co">#&gt; Calculating cluster 1</span></pre></body></html></div>
</div>
<div id="perform-pathway-enrichment-on-marker-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#perform-pathway-enrichment-on-marker-genes" class="anchor"></a>Perform pathway enrichment on marker genes</h3>
<p>Using the previously identified marker genes, we can perform a pathway enrichmeny analysis using the <code><a href="../reference/getEnrichedPathways.html">getEnrichedPathways()</a></code> function. The function will internally use the <a href="https://amp.pharm.mssm.edu/Enrichr/">Enrichr</a> API and store the results in the Seurat object under the <code>@misc</code> slot.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="no">seurat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/getEnrichedPathways.html">getEnrichedPathways</a></span>(
  <span class="no">seurat</span>,
  <span class="kw">column_sample</span> <span class="kw">=</span> <span class="st">'sample'</span>,
  <span class="kw">column_cluster</span> <span class="kw">=</span> <span class="st">'seurat_clusters'</span>,
  <span class="kw">adj_p_cutoff</span> <span class="kw">=</span> <span class="fl">0.01</span>,
  <span class="kw">max_terms</span> <span class="kw">=</span> <span class="fl">100</span>
)
<span class="co">#&gt; [16:13:19] Skipping pathway enrichment for samples because no marker genes were identified for any sample.</span>
<span class="co">#&gt; [16:13:19] Get enriched pathway for clusters...</span>
<span class="co">#&gt; [16:13:35] 16 pathways passed the thresholds across all clusters and databases.</span></pre></body></html></div>
</div>
<div id="gene-set-enrichment" class="section level3">
<h3 class="hasAnchor">
<a href="#gene-set-enrichment" class="anchor"></a>Gene set enrichment</h3>
<p>On top of querying the databases of the <a href="https://amp.pharm.mssm.edu/Enrichr/">Enrichr</a> service, one can also perform gene set enrichment analysis on gene sets provided as a <code>.gmt</code> file, e.g. from the <a href="https://www.gsea-msigdb.org/gsea/msigdb/index.jsp">MSigDB</a>. The <code><a href="../reference/performGeneSetEnrichmentAnalysis.html">performGeneSetEnrichmentAnalysis()</a></code> function uses the <a href="https://bioconductor.org/packages/release/bioc/html/GSVA.html">GSVA</a> method in combination with additional statistics as published by <a href="https://f1000research.com/articles/8-296">Diaz-Mejia et. al.</a>.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">example_gene_set</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata/example_gene_set.gmt"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"cerebroApp"</span>)

<span class="no">seurat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/performGeneSetEnrichmentAnalysis.html">performGeneSetEnrichmentAnalysis</a></span>(
  <span class="no">seurat</span>,
  <span class="kw">GMT_file</span> <span class="kw">=</span> <span class="no">example_gene_set</span>,
  <span class="kw">column_sample</span> <span class="kw">=</span> <span class="st">'sample'</span>,
  <span class="kw">column_cluster</span> <span class="kw">=</span> <span class="st">'seurat_clusters'</span>,
  <span class="kw">thresh_p_val</span> <span class="kw">=</span> <span class="fl">0.05</span>,
  <span class="kw">thresh_q_val</span> <span class="kw">=</span> <span class="fl">0.1</span>,
  <span class="kw">parallel.sz</span> <span class="kw">=</span> <span class="fl">1</span>,
  <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>
)</pre></body></html></div>
</div>
</div>
<div id="trajectory-analysis-with-monocle-optional" class="section level2">
<h2 class="hasAnchor">
<a href="#trajectory-analysis-with-monocle-optional" class="anchor"></a>Trajectory analysis with Monocle (optional)</h2>
<p>Then, we perform trajectory analysis with <a href="http://cole-trapnell-lab.github.io/monocle-release/">Monocle v2</a> using the previously identified highly variable genes. We extract the trajectory from the generated Monocle object with the <code><a href="../reference/extractMonocleTrajectory.html">extractMonocleTrajectory()</a></code> function of cerebroApp and attach it to our Seurat object.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="no">monocle</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle/man/newCellDataSet.html">newCellDataSet</a></span>(
  <span class="no">seurat</span>@<span class="kw">assays</span>$<span class="no">RNA</span>@<span class="kw">data</span>,
  <span class="kw">phenoData</span> <span class="kw">=</span> <span class="fu">new</span>(<span class="st">'AnnotatedDataFrame'</span>, <span class="kw">data</span> <span class="kw">=</span> <span class="no">seurat</span>@<span class="kw">meta.data</span>),
  <span class="kw">featureData</span> <span class="kw">=</span> <span class="fu">new</span>(<span class="st">'AnnotatedDataFrame'</span>, <span class="kw">data</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(
    <span class="kw">gene_short_name</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">seurat</span>@<span class="kw">assays</span>$<span class="no">RNA</span>@<span class="kw">counts</span>),
    <span class="kw">row.names</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">seurat</span>@<span class="kw">assays</span>$<span class="no">RNA</span>@<span class="kw">counts</span>))
  )
)

<span class="no">monocle</span> <span class="kw">&lt;-</span> <span class="fu">estimateSizeFactors</span>(<span class="no">monocle</span>)
<span class="no">monocle</span> <span class="kw">&lt;-</span> <span class="fu">estimateDispersions</span>(<span class="no">monocle</span>)
<span class="co">#&gt; Removing 9 outliers</span>
<span class="no">monocle</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle/man/setOrderingFilter.html">setOrderingFilter</a></span>(<span class="no">monocle</span>, <span class="no">seurat</span>@<span class="kw">assays</span>$<span class="no">RNA</span>@<span class="kw">var.features</span>)
<span class="no">monocle</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle/man/reduceDimension.html">reduceDimension</a></span>(<span class="no">monocle</span>, <span class="kw">max_components</span> <span class="kw">=</span> <span class="fl">2</span>, <span class="kw">method</span> <span class="kw">=</span> <span class="st">'DDRTree'</span>)
<span class="no">monocle</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle/man/orderCells.html">orderCells</a></span>(<span class="no">monocle</span>)

<span class="no">seurat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/extractMonocleTrajectory.html">extractMonocleTrajectory</a></span>(<span class="no">monocle</span>, <span class="no">seurat</span>, <span class="st">'highly_variable_genes'</span>)</pre></body></html></div>
</div>
<div id="export-to-cerebro-format" class="section level2">
<h2 class="hasAnchor">
<a href="#export-to-cerebro-format" class="anchor"></a>Export to Cerebro format</h2>
<p>Finally, we use the <code><a href="../reference/exportFromSeurat.html">exportFromSeurat()</a></code> function of cerebroApp to export our Seurat object to a <code>.crb</code> file which can be loaded into Cerebro.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="fu"><a href="../reference/exportFromSeurat.html">exportFromSeurat</a></span>(
  <span class="no">seurat</span>,
  <span class="kw">experiment_name</span> <span class="kw">=</span> <span class="st">'pbmc'</span>,
  <span class="kw">file</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">'cerebro_pbmc_'</span>, <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html">Sys.Date</a></span>(), <span class="st">'.crb'</span>),
  <span class="kw">organism</span> <span class="kw">=</span> <span class="st">'hg'</span>,
  <span class="kw">column_cluster</span> <span class="kw">=</span> <span class="st">'seurat_clusters'</span>,
  <span class="kw">column_nUMI</span> <span class="kw">=</span> <span class="st">'nCount_RNA'</span>,
  <span class="kw">column_nGene</span> <span class="kw">=</span> <span class="st">'nFeature_RNA'</span>
)
<span class="co">#&gt; [16:13:38] Initializing Cerebro object...</span>
<span class="co">#&gt; [16:13:38] Collecting available meta data...</span>
<span class="co">#&gt; [16:13:38] Extracting all meta data columns...</span>
<span class="co">#&gt; [16:13:38] Extracting tables of most expressed genes...</span>
<span class="co">#&gt; [16:13:38] Extracting tables of marker genes...</span>
<span class="co">#&gt; [16:13:38] Extracting pathway enrichment results...</span>
<span class="co">#&gt; [16:13:38] Extracting dimensional reductions...</span>
<span class="co">#&gt; [16:13:38] Will export the following dimensional reductions: UMAP, UMAP_3D</span>
<span class="co">#&gt; [16:13:38] Will export the following trajectories: highly_variable_genes</span>
<span class="co">#&gt; [16:13:38] Extracting cluster tree...</span>
<span class="co">#&gt; [16:13:38] Extracting log-normalized expression data...</span>
<span class="co">#&gt; [16:13:38] Saving Cerebro file...</span></pre></body></html></div>
<p>The Cerebro use interface can be launched using the <code><a href="../reference/launchCerebro.html">launchCerebro()</a></code> function.</p>
</div>
</div>
<div id="other-example-data-sets" class="section level1">
<h1 class="hasAnchor">
<a href="#other-example-data-sets" class="anchor"></a>Other example data sets</h1>
<p>We provide ready-to-load <code>.crb</code> files for a set of real data sets processed using either <a href="https://icb-scanpy.readthedocs-hosted.com/en/stable/">scanpy</a> and <a href="https://satijalab.org/seurat/">Seurat (v2.x and v3.x)</a> on our GitHub repository: <a href="https://github.com/romanhaa/Cerebro/tree/master/examples" class="uri">https://github.com/romanhaa/Cerebro/tree/master/examples</a></p>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
<span class="co">#&gt; R version 3.6.3 (2020-02-29)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin15.6.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Catalina 10.15.5</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt;  [1] splines   stats4    parallel  stats     graphics  grDevices utils    </span>
<span class="co">#&gt;  [8] datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] devtools_2.2.2      usethis_1.5.1       cerebroApp_1.2.2   </span>
<span class="co">#&gt;  [4] monocle_2.14.0      DDRTree_0.1.5       irlba_2.3.3        </span>
<span class="co">#&gt;  [7] VGAM_1.1-3          ggplot2_3.3.0       Biobase_2.46.0     </span>
<span class="co">#&gt; [10] BiocGenerics_0.32.0 Matrix_1.2-18       Seurat_3.1.4       </span>
<span class="co">#&gt; [13] dplyr_0.8.5         BiocManager_1.30.10</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] shinydashboard_0.7.1 reticulate_1.13      tidyselect_1.0.0    </span>
<span class="co">#&gt;   [4] RSQLite_2.2.0        AnnotationDbi_1.48.0 htmlwidgets_1.5.1   </span>
<span class="co">#&gt;   [7] grid_3.6.3           combinat_0.0-8       docopt_0.6.1        </span>
<span class="co">#&gt;  [10] Rtsne_0.15           munsell_0.5.0        codetools_0.2-16    </span>
<span class="co">#&gt;  [13] mutoss_0.1-12        ica_1.0-2            DT_0.12             </span>
<span class="co">#&gt;  [16] future_1.16.0        miniUI_0.1.1.1       withr_2.1.2         </span>
<span class="co">#&gt;  [19] colorspace_1.4-1     fastICA_1.2-2        knitr_1.28          </span>
<span class="co">#&gt;  [22] ROCR_1.0-7           shinyWidgets_0.5.1   gbRd_0.4-11         </span>
<span class="co">#&gt;  [25] listenv_0.8.0        Rdpack_0.11-1        slam_0.1-47         </span>
<span class="co">#&gt;  [28] mnormt_1.5-6         bit64_0.9-7          pheatmap_1.0.12     </span>
<span class="co">#&gt;  [31] rprojroot_1.3-2      vctrs_0.2.3          treeio_1.10.0       </span>
<span class="co">#&gt;  [34] TH.data_1.0-10       xfun_0.12            BiocFileCache_1.10.2</span>
<span class="co">#&gt;  [37] R6_2.4.1             rsvd_1.0.3           msigdbr_7.0.1       </span>
<span class="co">#&gt;  [40] bitops_1.0-6         assertthat_0.2.1     promises_1.1.0      </span>
<span class="co">#&gt;  [43] scales_1.1.0         multcomp_1.4-12      gtable_0.3.0        </span>
<span class="co">#&gt;  [46] npsurv_0.4-0         globals_0.12.5       processx_3.4.2      </span>
<span class="co">#&gt;  [49] sandwich_2.5-1       rlang_0.4.5          lazyeval_0.2.2      </span>
<span class="co">#&gt;  [52] formattable_0.2.0.1  yaml_2.2.1           reshape2_1.4.3      </span>
<span class="co">#&gt;  [55] backports_1.1.5      httpuv_1.5.2         qvalue_2.18.0       </span>
<span class="co">#&gt;  [58] tools_3.6.3          ellipsis_0.3.0       gplots_3.0.3        </span>
<span class="co">#&gt;  [61] RColorBrewer_1.1-2   proxy_0.4-24         sessioninfo_1.1.1   </span>
<span class="co">#&gt;  [64] ggridges_0.5.2       TFisher_0.2.0        Rcpp_1.0.3          </span>
<span class="co">#&gt;  [67] plyr_1.8.6           progress_1.2.2       purrr_0.3.3         </span>
<span class="co">#&gt;  [70] RCurl_1.98-1.1       densityClust_0.3     prettyunits_1.1.1   </span>
<span class="co">#&gt;  [73] ps_1.3.2             openssl_1.4.1        pbapply_1.4-2       </span>
<span class="co">#&gt;  [76] viridis_0.5.1        cowplot_1.0.0        S4Vectors_0.24.3    </span>
<span class="co">#&gt;  [79] zoo_1.8-7            ggrepel_0.9.0        cluster_2.1.0       </span>
<span class="co">#&gt;  [82] fs_1.3.2             magrittr_1.5         RSpectra_0.16-0     </span>
<span class="co">#&gt;  [85] data.table_1.12.8    colourpicker_1.0     lmtest_0.9-37       </span>
<span class="co">#&gt;  [88] RANN_2.6.1           mvtnorm_1.1-0        fitdistrplus_1.0-14 </span>
<span class="co">#&gt;  [91] matrixStats_0.55.0   pkgload_1.0.2        hms_0.5.3           </span>
<span class="co">#&gt;  [94] patchwork_1.0.0      lsei_1.2-0           mime_0.9            </span>
<span class="co">#&gt;  [97] evaluate_0.14        GSVA_1.34.0          xtable_1.8-4        </span>
<span class="co">#&gt; [100] XML_3.99-0.3         sparsesvd_0.2        IRanges_2.20.2      </span>
<span class="co">#&gt; [103] gridExtra_2.3        biomaRt_2.42.0       testthat_2.3.2      </span>
<span class="co">#&gt; [106] HSMMSingleCell_1.6.0 compiler_3.6.3       tibble_2.1.3        </span>
<span class="co">#&gt; [109] KernSmooth_2.23-16   crayon_1.3.4         htmltools_0.4.0     </span>
<span class="co">#&gt; [112] later_1.0.0          tidyr_1.0.2          geneplotter_1.64.0  </span>
<span class="co">#&gt; [115] RcppParallel_4.4.4   DBI_1.1.0            dbplyr_1.4.2        </span>
<span class="co">#&gt; [118] rappdirs_0.3.1       MASS_7.3-51.5        readr_1.3.1         </span>
<span class="co">#&gt; [121] cli_2.0.2            gdata_2.18.0         metap_1.3           </span>
<span class="co">#&gt; [124] igraph_1.2.4.2       pkgconfig_2.0.3      pkgdown_1.5.1       </span>
<span class="co">#&gt; [127] sn_1.5-5             rvcheck_0.1.8        numDeriv_2016.8-1.1 </span>
<span class="co">#&gt; [130] plotly_4.9.2         ggtree_2.0.2         annotate_1.64.0     </span>
<span class="co">#&gt; [133] multtest_2.42.0      bibtex_0.4.2.2       callr_3.4.2         </span>
<span class="co">#&gt; [136] stringr_1.4.0        digest_0.6.25        sctransform_0.2.1   </span>
<span class="co">#&gt; [139] RcppAnnoy_0.0.15     tsne_0.1-3           graph_1.64.0        </span>
<span class="co">#&gt; [142] rmarkdown_2.1        leiden_0.3.3         tidytree_0.3.1      </span>
<span class="co">#&gt; [145] uwot_0.1.5           GSEABase_1.48.0      curl_4.3            </span>
<span class="co">#&gt; [148] shiny_1.4.0          gtools_3.8.1         lifecycle_0.2.0     </span>
<span class="co">#&gt; [151] nlme_3.1-144         jsonlite_1.6.1       askpass_1.1         </span>
<span class="co">#&gt; [154] fansi_0.4.1          desc_1.2.0           viridisLite_0.3.0   </span>
<span class="co">#&gt; [157] limma_3.42.2         pillar_1.4.3         lattice_0.20-38     </span>
<span class="co">#&gt; [160] pkgbuild_1.0.6       fastmap_1.0.1        httr_1.4.1          </span>
<span class="co">#&gt; [163] plotrix_3.7-7        survival_3.1-8       remotes_2.1.1       </span>
<span class="co">#&gt; [166] glue_1.3.1           qlcMatrix_0.9.7      FNN_1.1.3           </span>
<span class="co">#&gt; [169] png_0.1-7            shinythemes_1.1.2    bit_1.1-15.2        </span>
<span class="co">#&gt; [172] stringi_1.4.6        blob_1.2.1           caTools_1.18.0      </span>
<span class="co">#&gt; [175] memoise_1.1.0        future.apply_1.4.0   ape_5.3</span></pre></body></html></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Roman Hillje.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
