<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Export a data set in SCE format • cerebroApp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Export a data set in SCE format">
<meta property="og:description" content="cerebroApp">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cerebroApp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cerebroApp_workflow_Seurat.html">The cerebroApp workflow (Seurat)</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">The Cerebro interface</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/Cerebro_interface_v1.2.html">Cerebro v1.2 and older</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../articles/export_a_data_set_in_SCE_format.html">Export an SCE/SingleCellExperiment object</a>
    </li>
    <li>
      <a href="../articles/export_and_visualize_custom_tables_and_plots.html">Export and visualize custom tables and plots</a>
    </li>
    <li>
      <a href="../articles/launch_cerebro_with_pre-loaded_data_set.html">Launch Cerebro with pre-loaded data set</a>
    </li>
    <li>
      <a href="../articles/host_cerebro_on_shinyapps.html">Host Cerebro on shinyapps.io</a>
    </li>
    <li>
      <a href="../articles/control_access_to_cerebro_with_a_login_page.html">Control access to Cerebro with a login page</a>
    </li>
    <li>
      <a href="../articles/overview_of_cerebro_v1.3_class.html">Overview of the Cerebro v1.3 class</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Release notes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/release_notes_v1.3.html">v1.3</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/romanhaa/cerebroApp/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/fakechek1">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="export_a_data_set_in_SCE_format_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Export a data set in SCE format</h1>
                        <h4 class="author">Roman Hillje</h4>
            
            <h4 class="date">25 September, 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/romanhaa/cerebroApp/blob/master/vignettes/export_a_data_set_in_SCE_format.Rmd"><code>vignettes/export_a_data_set_in_SCE_format.Rmd</code></a></small>
      <div class="hidden name"><code>export_a_data_set_in_SCE_format.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<p>Here, I explain how one can export a data set stored in the <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html"><code>SingleCellExperiment</code>/<code>SCE</code></a> format to visualize it in Cerebro.</p>
<p>The workflow used here is loosely put together from the <a href="https://bioconductor.org/packages/release/bioc/vignettes/scran/inst/doc/scran.html"><code>scran</code></a> and <a href="https://bioconductor.org/packages/3.11/bioc/vignettes/scater/inst/doc/overview.html"><code>scater</code></a> examples. Do not use this workflow as a reference for how to properly process your data.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>Before starting the workflow, we need to install <code>cerebroApp</code>, as well as the <a href="https://bioconductor.org/packages/release/bioc/html/scran.html"><code>scran</code></a>, <a href="https://bioconductor.org/packages/3.11/bioc/html/scater.html"><code>scater</code></a> and <a href="https://www.bioconductor.org/packages/release/bioc/html/SingleR.html"><code>SingleR</code></a> packages, which are not installed as dependencies of cerebroApp because they are only necessary if you want/need to pre-process your scRNA-seq data. If you just want to launch the Cerebro user interface, e.g. because you already have the pre-processed data, you don’t need those packages.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="kw">if</span> ( <span class="st">'BiocManager'</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span>() <span class="kw">==</span> <span class="fl">FALSE</span> ) <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">'BiocManager'</span>)

<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BiocManager</span>)

<span class="kw">if</span> ( <span class="st">'cerebroApp'</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span>() <span class="kw">==</span> <span class="fl">FALSE</span> ) <span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">'cerebroApp'</span>)
<span class="kw">if</span> ( <span class="st">'scran'</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span>() <span class="kw">==</span> <span class="fl">FALSE</span> ) <span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">'scran'</span>)
<span class="kw">if</span> ( <span class="st">'scater'</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span>() <span class="kw">==</span> <span class="fl">FALSE</span> ) <span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">'scater'</span>)
<span class="kw">if</span> ( <span class="st">'SingleR'</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span>() <span class="kw">==</span> <span class="fl">FALSE</span> ) <span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">'SingleR'</span>)
<span class="kw">if</span> ( <span class="st">'monocle'</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span>() <span class="kw">==</span> <span class="fl">FALSE</span> ) <span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">'monocle'</span>)</pre></body></html></div>
</div>
<div id="load-packages" class="section level1">
<h1 class="hasAnchor">
<a href="#load-packages" class="anchor"></a>Load packages</h1>
<p>Apart from the packages we just installed, we will also load the <code>dplyr</code> package and set a seed to make our analysis reproducible.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dplyr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">scran</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">scater</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">SingleR</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">cerebroApp</span>)

<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="kw">width</span> <span class="kw">=</span> <span class="fl">100</span>)
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1234567</span>)</pre></body></html></div>
</div>
<div id="load-count-data" class="section level1">
<h1 class="hasAnchor">
<a href="#load-count-data" class="anchor"></a>Load count data</h1>
<p>In this example workflow, we will load a small transcript count table from the <a href="https://satijalab.org/seurat/"><code>Seurat</code></a> package containing 80 cells and 230 genes. If you want to try the workflow with a real data set, you can find some transcript count matrices on the <a href="https://github.com/romanhaa/Cerebro/tree/master/examples">Cerebro GitHub repo</a>.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">pbmc_counts</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(
    <span class="kw">file</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'pbmc_raw.txt'</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">'Seurat'</span>),
    <span class="kw">as.is</span> <span class="kw">=</span> <span class="fl">TRUE</span>
  ) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()</pre></body></html></div>
</div>
<div id="pre-processing-with-scran-and-scater" class="section level1">
<h1 class="hasAnchor">
<a href="#pre-processing-with-scran-and-scater" class="anchor"></a>Pre-processing with <code>scran</code> and <code>scater</code>
</h1>
<p>Now, we create a <code>SingleCellExperiment</code> object and filter out cells with less than <code>50</code> transcripts or fewer than <code>10</code> expressed genes. Those cut-offs are only reasonable for this example data set and will likely need to be adjusted in a real data set. Then, we…</p>
<ul>
<li>remove genes expressed in less than 5 cells,</li>
<li>log-normalize transcript counts,</li>
<li>identify highly variably genes,</li>
<li>perform principal component analysis (PCA).</li>
</ul>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">sce</span> <span class="kw">&lt;-</span> <span class="kw pkg">SingleCellExperiment</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html">SingleCellExperiment</a></span>(
  <span class="kw">assays</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">counts</span> <span class="kw">=</span> <span class="no">pbmc_counts</span>)
)

<span class="no">qc_metrics</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/perCellQCMetrics.html">perCellQCMetrics</a></span>(<span class="no">sce</span>)
<span class="fu">colData</span>(<span class="no">sce</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="fu">colData</span>(<span class="no">sce</span>), <span class="no">qc_metrics</span>)

<span class="no">cells_to_keep</span> <span class="kw">&lt;-</span> <span class="no">sce</span>$<span class="no">sum</span> <span class="kw">&gt;=</span> <span class="fl">50</span> <span class="kw">&amp;</span> <span class="no">sce</span>$<span class="no">detected</span> <span class="kw">&gt;=</span> <span class="fl">10</span>
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="no">sce</span>[,<span class="no">cells_to_keep</span>]

<span class="no">genes_to_keep</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/nexprs.html">nexprs</a></span>(<span class="no">sce</span>, <span class="kw">byrow</span><span class="kw">=</span><span class="fl">TRUE</span>) <span class="kw">&gt;=</span> <span class="fl">5</span>
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="no">sce</span>[<span class="no">genes_to_keep</span>,]

<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/logNormCounts.html">logNormCounts</a></span>(<span class="no">sce</span>)

<span class="no">variable_genes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html">modelGeneVar</a></span>(<span class="no">sce</span>)
<span class="no">highly_variable_genes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html">getTopHVGs</a></span>(<span class="no">variable_genes</span>, <span class="kw">n</span> <span class="kw">=</span> <span class="fl">50</span>)

<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runPCA.html">runPCA</a></span>(<span class="no">sce</span>)</pre></body></html></div>
</div>
<div id="clustering" class="section level1">
<h1 class="hasAnchor">
<a href="#clustering" class="anchor"></a>Clustering</h1>
<p>Then, we identify clusters using a graph-based approach.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">graph</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html">buildSNNGraph</a></span>(<span class="no">sce</span>, <span class="kw">use.dimred</span> <span class="kw">=</span> <span class="st">"PCA"</span>)
<span class="no">cluster</span> <span class="kw">&lt;-</span> <span class="kw pkg">igraph</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/cluster_walktrap.html">cluster_walktrap</a></span>(<span class="no">graph</span>)$<span class="no">membership</span>
<span class="no">sce</span>$<span class="no">cluster</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">cluster</span>)</pre></body></html></div>
</div>
<div id="dimensional-reduction" class="section level1">
<h1 class="hasAnchor">
<a href="#dimensional-reduction" class="anchor"></a>Dimensional reduction</h1>
<p>Next, we use the UMAP technique to generate two- and three-dimensional projections.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span>(
  <span class="no">sce</span>,
  <span class="kw">name</span> <span class="kw">=</span> <span class="st">'UMAP'</span>,
  <span class="kw">ncomponents</span> <span class="kw">=</span> <span class="fl">2</span>,
)

<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span>(
  <span class="no">sce</span>,
  <span class="kw">name</span> <span class="kw">=</span> <span class="st">'UMAP_3D'</span>,
  <span class="kw">ncomponents</span> <span class="kw">=</span> <span class="fl">3</span>,
)</pre></body></html></div>
</div>
<div id="curate-meta-data" class="section level1">
<h1 class="hasAnchor">
<a href="#curate-meta-data" class="anchor"></a>Curate meta data</h1>
<p>This example data set consists of a single sample so we just add that name to the meta data. Moreover, in order to be able to understand how we did the analysis later, we add some meta data to the <code>misc</code> slot of our <code>SCE</code> object, e.g. an experiment name and the species. The info we store there will later be collected for visualization in Cerebro.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">sce</span>$<span class="no">sample</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="st">'pbmc'</span>, <span class="kw">levels</span> <span class="kw">=</span> <span class="st">'pbmc'</span>)

<span class="no">sce</span>@<span class="kw">metadata</span>$<span class="no">experiment</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(
  <span class="kw">experiment_name</span> <span class="kw">=</span> <span class="st">'pbmc'</span>,
  <span class="kw">organism</span> <span class="kw">=</span> <span class="st">'hg'</span>,
  <span class="kw">date_of_analysis</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html">Sys.Date</a></span>()
)

<span class="no">sce</span>@<span class="kw">metadata</span>$<span class="no">parameters</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(
  <span class="kw">gene_nomenclature</span> <span class="kw">=</span> <span class="st">'gene_name'</span>,
  <span class="kw">discard_genes_expressed_in_fewer_cells_than</span> <span class="kw">=</span> <span class="fl">10</span>,
  <span class="kw">keep_mitochondrial_genes</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
  <span class="kw">variables_to_regress_out</span> <span class="kw">=</span> <span class="st">'nUMI'</span>
)

<span class="no">sce</span>@<span class="kw">metadata</span>$<span class="no">parameters</span>$<span class="no">filtering</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(
  <span class="kw">UMI_min</span> <span class="kw">=</span> <span class="fl">50</span>,
  <span class="kw">UMI_max</span> <span class="kw">=</span> <span class="fl">Inf</span>,
  <span class="kw">genes_min</span> <span class="kw">=</span> <span class="fl">10</span>,
  <span class="kw">genes_max</span> <span class="kw">=</span> <span class="fl">Inf</span>
)

<span class="no">sce</span>@<span class="kw">metadata</span>$<span class="no">technical_info</span>$<span class="no">cerebroApp_version</span> <span class="kw">&lt;-</span> <span class="kw pkg">utils</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html">packageVersion</a></span>(<span class="st">'cerebroApp'</span>)
<span class="no">sce</span>@<span class="kw">metadata</span>$<span class="no">technical_info</span>$<span class="no">scran</span> <span class="kw">&lt;-</span> <span class="kw pkg">utils</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html">packageVersion</a></span>(<span class="st">'scran'</span>)
<span class="no">sce</span>@<span class="kw">metadata</span>$<span class="no">technical_info</span>$<span class="no">scater</span> <span class="kw">&lt;-</span> <span class="kw pkg">utils</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html">packageVersion</a></span>(<span class="st">'scater'</span>)
<span class="no">sce</span>@<span class="kw">metadata</span>$<span class="no">technical_info</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(
  <span class="st">'R'</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/capture.output.html">capture.output</a></span>(<span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/reexports.html">session_info</a></span>())
)</pre></body></html></div>
</div>
<div id="assign-cell-types" class="section level1">
<h1 class="hasAnchor">
<a href="#assign-cell-types" class="anchor"></a>Assign cell types</h1>
<p>Using the <a href="https://www.bioconductor.org/packages/release/bioc/html/SingleR.html"><code>SingleR</code></a> package, we can get a suggestion of cell type for each cell. Apart from the suggested cell type, we also extract the score that might help us to understand how confident the assignment was.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">singler_ref</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/BlueprintEncodeData.html">BlueprintEncodeData</a></span>()
<span class="co">#&gt; Warning: `select_()` is deprecated as of dplyr 0.7.0.</span>
<span class="co">#&gt; Please use `select()` instead.</span>
<span class="co">#&gt; This warning is displayed once every 8 hours.</span>
<span class="co">#&gt; Call `lifecycle::last_warnings()` to see where this warning was generated.</span>
<span class="co">#&gt; Warning: `filter_()` is deprecated as of dplyr 0.7.0.</span>
<span class="co">#&gt; Please use `filter()` instead.</span>
<span class="co">#&gt; See vignette('programming') for more help</span>
<span class="co">#&gt; This warning is displayed once every 8 hours.</span>
<span class="co">#&gt; Call `lifecycle::last_warnings()` to see where this warning was generated.</span>
<span class="co">#&gt; snapshotDate(): 2019-10-22</span>
<span class="co">#&gt; see ?SingleR and browseVignettes('SingleR') for documentation</span>
<span class="co">#&gt; loading from cache</span>
<span class="co">#&gt; see ?SingleR and browseVignettes('SingleR') for documentation</span>
<span class="co">#&gt; loading from cache</span>

<span class="no">singler_results_blueprintencode_main</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html">SingleR</a></span>(
  <span class="kw">test</span> <span class="kw">=</span> <span class="fu">logcounts</span>(<span class="no">sce</span>),
  <span class="kw">ref</span> <span class="kw">=</span> <span class="no">singler_ref</span>,
  <span class="kw">labels</span> <span class="kw">=</span> <span class="no">singler_ref</span>@<span class="kw">colData</span>@<span class="kw">listData</span>$<span class="no">label.main</span>
)

<span class="no">sce</span>$<span class="no">cell_type_singler_blueprintencode_main</span> <span class="kw">&lt;-</span> <span class="no">singler_results_blueprintencode_main</span>@<span class="kw">listData</span>$<span class="no">labels</span>

<span class="no">singler_scores</span> <span class="kw">&lt;-</span> <span class="no">singler_results_blueprintencode_main</span>@<span class="kw">listData</span>$<span class="no">scores</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/reexports.html">as_tibble</a></span>() <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">assigned_score</span> <span class="kw">=</span> <span class="fl">NA</span>)

<span class="kw">for</span> ( <span class="no">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">singler_scores</span>)) ) {
  <span class="no">singler_scores</span>$<span class="no">assigned_score</span>[<span class="no">i</span>] <span class="kw">&lt;-</span> <span class="no">singler_scores</span><span class="kw">[[</span><span class="no">singler_results_blueprintencode_main</span>@<span class="kw">listData</span>$<span class="no">labels</span>[<span class="no">i</span>]]][<span class="no">i</span>]
}

<span class="no">sce</span>$<span class="no">cell_type_singler_blueprintencode_main_score</span> <span class="kw">&lt;-</span> <span class="no">singler_scores</span>$<span class="no">assigned_score</span></pre></body></html></div>
<p>##Identify marker genes</p>
<p>Marker genes are genes which are particularly strongly or weakly expressed in a given cell population, e.g. a cluster. This information can help to distinguish the role of different cell groups in the data set. Here, marker genes are identified for clusters and cell types using the <code><a href="https://rdrr.io/pkg/scran/man/findMarkers.html">findMarkers()</a></code> function from the <code>scran</code> package. The results are separate tables for the respective subgroups, which we merge together in order to make them compatible with Cerebro.</p>
<div id="clusters" class="section level2">
<h2 class="hasAnchor">
<a href="#clusters" class="anchor"></a>Clusters</h2>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">sce</span>@<span class="kw">metadata</span>$<span class="no">marker_genes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>()
<span class="no">sce</span>@<span class="kw">metadata</span>$<span class="no">marker_genes</span>$<span class="no">scran</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>()

<span class="no">marker_genes_cluster</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html">findMarkers</a></span>(<span class="no">sce</span>, <span class="no">sce</span>$<span class="no">cluster</span>)

<span class="kw">for</span> ( <span class="no">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span>(<span class="no">marker_genes_cluster</span>) ) {
  <span class="no">marker_genes_cluster</span><span class="kw">[[</span><span class="no">i</span>]] <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="no">marker_genes_cluster</span><span class="kw">[[</span><span class="no">i</span>]]) <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(
      <span class="kw">cluster</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">marker_genes_cluster</span>)[<span class="no">i</span>],
      <span class="kw">gene</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">.</span>)
    )
  <span class="no">marker_genes_cluster</span><span class="kw">[[</span><span class="no">i</span>]]<span class="kw">[[</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">'logFC.'</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">marker_genes_cluster</span>)[<span class="no">i</span>])]] <span class="kw">&lt;-</span> <span class="fl">0</span>
}

<span class="no">marker_genes_cluster</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="no">bind_rows</span>, <span class="no">marker_genes_cluster</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="no">cluster</span>, <span class="no">gene</span>, <span class="no">Top</span>, <span class="no">p.value</span>, <span class="no">FDR</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/reexports.html">everything</a></span>()) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">cluster</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">cluster</span>, <span class="kw">levels</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span>(<span class="fu">colData</span>(<span class="no">sce</span>)$<span class="no">cluster</span>)))

<span class="no">sce</span>@<span class="kw">metadata</span>$<span class="no">marker_genes</span>$<span class="no">scran</span>$<span class="no">cluster</span> <span class="kw">&lt;-</span> <span class="no">marker_genes_cluster</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(!<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">Top</span>), <span class="no">FDR</span> <span class="kw">&lt;=</span> <span class="fl">0.1</span>)

<span class="fu"><a href="https://dplyr.tidyverse.org/reference/reexports.html">glimpse</a></span>(<span class="no">marker_genes_cluster</span>)
<span class="co">#&gt; Rows: 876</span>
<span class="co">#&gt; Columns: 9</span>
<span class="co">#&gt; $ cluster &lt;fct&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, …</span>
<span class="co">#&gt; $ gene    &lt;chr&gt; "HLA-DRA", "CST3", "GPX1", "TYMP", "HLA-DRB1", "PPBP", "CD79B", "LYZ", "IFITM2", …</span>
<span class="co">#&gt; $ Top     &lt;int&gt; 1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 4, 5, 5, 6, 6, 7, 7, 8, 8, 8, 9, 9, 10, 11, 11, …</span>
<span class="co">#&gt; $ p.value &lt;dbl&gt; 7.673681e-23, 3.683285e-30, 2.667130e-17, 1.174240e-17, 2.632735e-11, 1.904290e-1…</span>
<span class="co">#&gt; $ FDR     &lt;dbl&gt; 8.402681e-21, 8.066393e-28, 1.168203e-15, 7.557534e-16, 7.207111e-10, 5.957706e-1…</span>
<span class="co">#&gt; $ logFC.2 &lt;dbl&gt; -2.60629181, -3.68112226, -1.05420348, -1.67732667, -1.96732725, -0.04586014, -0.…</span>
<span class="co">#&gt; $ logFC.3 &lt;dbl&gt; -4.74673175, -0.12080297, -0.05119315, 0.13734308, -3.68688025, -0.05364607, -2.7…</span>
<span class="co">#&gt; $ logFC.4 &lt;dbl&gt; -0.09773581, -0.79739926, -3.54934421, 0.13734308, 0.28378634, -5.69442304, 0.236…</span>
<span class="co">#&gt; $ logFC.1 &lt;dbl&gt; 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, …</span>

<span class="no">marker_genes_cluster</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>() <span class="kw">%&gt;%</span> <span class="kw pkg">knitr</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span>()</pre></body></html></div>
<table class="table">
<thead><tr class="header">
<th align="left">cluster</th>
<th align="left">gene</th>
<th align="right">Top</th>
<th align="right">p.value</th>
<th align="right">FDR</th>
<th align="right">logFC.2</th>
<th align="right">logFC.3</th>
<th align="right">logFC.4</th>
<th align="right">logFC.1</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">1</td>
<td align="left">HLA-DRA</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">-2.6062918</td>
<td align="right">-4.7467317</td>
<td align="right">-0.0977358</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="left">CST3</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">-3.6811223</td>
<td align="right">-0.1208030</td>
<td align="right">-0.7973993</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">1</td>
<td align="left">GPX1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">-1.0542035</td>
<td align="right">-0.0511932</td>
<td align="right">-3.5493442</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="left">TYMP</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">-1.6773267</td>
<td align="right">0.1373431</td>
<td align="right">0.1373431</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">1</td>
<td align="left">HLA-DRB1</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">-1.9673272</td>
<td align="right">-3.6868802</td>
<td align="right">0.2837863</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="left">PPBP</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">-0.0458601</td>
<td align="right">-0.0536461</td>
<td align="right">-5.6944230</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
</div>
<div id="cell-types" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-types" class="anchor"></a>Cell types</h2>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">marker_genes_cell_type</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html">findMarkers</a></span>(<span class="no">sce</span>, <span class="no">sce</span>$<span class="no">cell_type_singler_blueprintencode_main</span>)
<span class="co">#&gt; Warning in .pairwise_blocked_template(x, clust.vals, nblocks, direction = direction, : no within-</span>
<span class="co">#&gt; block comparison between Fibroblasts and B-cells</span>
<span class="co">#&gt; Warning in .pairwise_blocked_template(x, clust.vals, nblocks, direction = direction, : no within-</span>
<span class="co">#&gt; block comparison between Fibroblasts and CD4+ T-cells</span>
<span class="co">#&gt; Warning in .pairwise_blocked_template(x, clust.vals, nblocks, direction = direction, : no within-</span>
<span class="co">#&gt; block comparison between Fibroblasts and CD8+ T-cells</span>
<span class="co">#&gt; Warning in .pairwise_blocked_template(x, clust.vals, nblocks, direction = direction, : no within-</span>
<span class="co">#&gt; block comparison between Fibroblasts and Endothelial cells</span>
<span class="co">#&gt; Warning in .pairwise_blocked_template(x, clust.vals, nblocks, direction = direction, : no within-</span>
<span class="co">#&gt; block comparison between Fibroblasts and Erythrocytes</span>
<span class="co">#&gt; Warning in .pairwise_blocked_template(x, clust.vals, nblocks, direction = direction, : no within-</span>
<span class="co">#&gt; block comparison between Monocytes and Fibroblasts</span>
<span class="co">#&gt; Warning in .pairwise_blocked_template(x, clust.vals, nblocks, direction = direction, : no within-</span>
<span class="co">#&gt; block comparison between NK cells and Fibroblasts</span>

<span class="kw">for</span> ( <span class="no">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span>(<span class="no">marker_genes_cell_type</span>) ) {
  <span class="no">marker_genes_cell_type</span><span class="kw">[[</span><span class="no">i</span>]] <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="no">marker_genes_cell_type</span><span class="kw">[[</span><span class="no">i</span>]]) <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(
      <span class="kw">cell_type_singler_blueprintencode_main</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">marker_genes_cell_type</span>)[<span class="no">i</span>],
      <span class="kw">gene</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">.</span>)
    )
  <span class="no">marker_genes_cell_type</span><span class="kw">[[</span><span class="no">i</span>]]<span class="kw">[[</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">'logFC.'</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">marker_genes_cell_type</span>)[<span class="no">i</span>], <span class="kw">pattern</span> <span class="kw">=</span> <span class="st">' |\\-|\\+'</span>, <span class="kw">replacement</span> <span class="kw">=</span> <span class="st">'.'</span>))]] <span class="kw">&lt;-</span> <span class="fl">0</span>
}

<span class="no">marker_genes_cell_type</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="no">bind_rows</span>, <span class="no">marker_genes_cell_type</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="no">cell_type_singler_blueprintencode_main</span>, <span class="no">gene</span>, <span class="no">Top</span>, <span class="no">p.value</span>, <span class="no">FDR</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/reexports.html">everything</a></span>()) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">cell_type_singler_blueprintencode_main</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(
      <span class="no">cell_type_singler_blueprintencode_main</span>,
      <span class="kw">levels</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">cell_type_singler_blueprintencode_main</span>)
    )
  )

<span class="no">sce</span>@<span class="kw">metadata</span>$<span class="no">marker_genes</span>$<span class="no">scran</span>$<span class="no">cell_type_singler_blueprintencode_main</span> <span class="kw">&lt;-</span> <span class="no">marker_genes_cell_type</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(!<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">Top</span>), <span class="no">FDR</span> <span class="kw">&lt;=</span> <span class="fl">0.1</span>)

<span class="fu"><a href="https://dplyr.tidyverse.org/reference/reexports.html">glimpse</a></span>(<span class="no">marker_genes_cell_type</span>)
<span class="co">#&gt; Rows: 1,752</span>
<span class="co">#&gt; Columns: 13</span>
<span class="co">#&gt; $ cell_type_singler_blueprintencode_main &lt;fct&gt; B-cells, B-cells, B-cells, B-cells, B-cells, B-cel…</span>
<span class="co">#&gt; $ gene                                   &lt;chr&gt; "HLA-DRA", "TYROBP", "HLA-DRB1", "PPBP", "MS4A1", …</span>
<span class="co">#&gt; $ Top                                    &lt;int&gt; 1, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 5, 5, 5, 5, 6,…</span>
<span class="co">#&gt; $ p.value                                &lt;dbl&gt; 3.945181e-12, 7.668231e-19, 2.381471e-09, 7.718397…</span>
<span class="co">#&gt; $ FDR                                    &lt;dbl&gt; 1.439991e-10, 1.679343e-16, 6.519278e-08, 4.225822…</span>
<span class="co">#&gt; $ logFC.CD4..T.cells                     &lt;dbl&gt; 4.42860022, -0.60150095, 3.97066659, 0.12054811, 3…</span>
<span class="co">#&gt; $ logFC.CD8..T.cells                     &lt;dbl&gt; 4.97694333, -0.44507730, 3.63219454, -0.09502514, …</span>
<span class="co">#&gt; $ logFC.Endothelial.cells                &lt;dbl&gt; 4.8046017, 0.0000000, 3.9706666, -6.0736140, 3.110…</span>
<span class="co">#&gt; $ logFC.Erythrocytes                     &lt;dbl&gt; 4.56757886, 0.00000000, 3.97066659, -5.67059909, 3…</span>
<span class="co">#&gt; $ logFC.Fibroblasts                      &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA…</span>
<span class="co">#&gt; $ logFC.Monocytes                        &lt;dbl&gt; 2.140439941, -2.981126027, 1.719552998, 0.00778592…</span>
<span class="co">#&gt; $ logFC.NK.cells                         &lt;dbl&gt; 4.78616075, -2.34779876, 3.53870363, 0.12054811, 3…</span>
<span class="co">#&gt; $ logFC.B.cells                          &lt;dbl&gt; 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,…</span>

<span class="no">marker_genes_cell_type</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="fl">1</span>:<span class="fl">6</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>() <span class="kw">%&gt;%</span> <span class="kw pkg">knitr</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span>(<span class="kw">booktabs</span> <span class="kw">=</span> <span class="no">T</span>)</pre></body></html></div>
<table class="table">
<colgroup>
<col width="45%">
<col width="10%">
<col width="4%">
<col width="9%">
<col width="7%">
<col width="22%">
</colgroup>
<thead><tr class="header">
<th align="left">cell_type_singler_blueprintencode_main</th>
<th align="left">gene</th>
<th align="right">Top</th>
<th align="right">p.value</th>
<th align="right">FDR</th>
<th align="right">logFC.CD4..T.cells</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">B-cells</td>
<td align="left">HLA-DRA</td>
<td align="right">1</td>
<td align="right">0e+00</td>
<td align="right">0e+00</td>
<td align="right">4.4286002</td>
</tr>
<tr class="even">
<td align="left">B-cells</td>
<td align="left">TYROBP</td>
<td align="right">1</td>
<td align="right">0e+00</td>
<td align="right">0e+00</td>
<td align="right">-0.6015010</td>
</tr>
<tr class="odd">
<td align="left">B-cells</td>
<td align="left">HLA-DRB1</td>
<td align="right">1</td>
<td align="right">0e+00</td>
<td align="right">1e-07</td>
<td align="right">3.9706666</td>
</tr>
<tr class="even">
<td align="left">B-cells</td>
<td align="left">PPBP</td>
<td align="right">1</td>
<td align="right">0e+00</td>
<td align="right">0e+00</td>
<td align="right">0.1205481</td>
</tr>
<tr class="odd">
<td align="left">B-cells</td>
<td align="left">MS4A1</td>
<td align="right">2</td>
<td align="right">3e-07</td>
<td align="right">3e-06</td>
<td align="right">3.1103836</td>
</tr>
<tr class="even">
<td align="left">B-cells</td>
<td align="left">HLA-DQB1</td>
<td align="right">2</td>
<td align="right">0e+00</td>
<td align="right">7e-07</td>
<td align="right">2.5485368</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="export-to-cerebro-format" class="section level1">
<h1 class="hasAnchor">
<a href="#export-to-cerebro-format" class="anchor"></a>Export to Cerebro format</h1>
<p>Finally, we use the <code><a href="../reference/exportFromSeurat.html">exportFromSeurat()</a></code> function of cerebroApp to export our Seurat object to a <code>.crb</code> file which can be loaded into Cerebro.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="fu"><a href="../reference/exportFromSCE.html">exportFromSCE</a></span>(
  <span class="no">sce</span>,
  <span class="kw">assay</span> <span class="kw">=</span> <span class="st">'logcounts'</span>,
  <span class="kw">file</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">'cerebro_pbmc_SCE_'</span>, <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html">Sys.Date</a></span>(), <span class="st">'.crb'</span>),
  <span class="kw">experiment_name</span> <span class="kw">=</span> <span class="st">'pbmc'</span>,
  <span class="kw">organism</span> <span class="kw">=</span> <span class="st">'hg'</span>,
  <span class="kw">groups</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'sample'</span>,<span class="st">'cluster'</span>,<span class="st">'cell_type_singler_blueprintencode_main'</span>),
  <span class="kw">nUMI</span> <span class="kw">=</span> <span class="st">'sum'</span>,
  <span class="kw">nGene</span> <span class="kw">=</span> <span class="st">'detected'</span>,
  <span class="kw">add_all_meta_data</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
  <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>
)
<span class="co">#&gt; [12:39:03] Start collecting data...</span>
<span class="co">#&gt; [12:39:03] Overview of Cerebro object:</span>
<span class="co">#&gt; class: Cerebro_v1.3</span>
<span class="co">#&gt; cerebroApp version: 1.3.0</span>
<span class="co">#&gt; experiment name: pbmc</span>
<span class="co">#&gt; organism: hg</span>
<span class="co">#&gt; date of analysis: 2020-09-25</span>
<span class="co">#&gt; date of export: 2020-09-25</span>
<span class="co">#&gt; number of cells: 79</span>
<span class="co">#&gt; number of genes: 219</span>
<span class="co">#&gt; grouping variables (3): sample, cluster, cell_type_singler_blueprintencode_main</span>
<span class="co">#&gt; cell cycle variables (0): </span>
<span class="co">#&gt; projections (2): UMAP, UMAP_3D</span>
<span class="co">#&gt; trees (0): </span>
<span class="co">#&gt; most expressed genes: </span>
<span class="co">#&gt; marker genes:</span>
<span class="co">#&gt;   - scran (2): cluster, cell_type_singler_blueprintencode_main</span>
<span class="co">#&gt; enriched pathways:</span>
<span class="co">#&gt; trajectories:</span>
<span class="co">#&gt; extra material:</span>
<span class="co">#&gt; [12:39:03] Saving Cerebro object to: cerebro_pbmc_SCE_2020-09-25.crb</span>
<span class="co">#&gt; [12:39:03] Done!</span></pre></body></html></div>
<p>The Cerebro use interface can be launched using the <code><a href="../reference/launchCerebro.html">launchCerebro()</a></code> function.</p>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
<span class="co">#&gt; R version 3.6.3 (2020-02-29)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin15.6.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Catalina 10.15.6</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] cerebroApp_1.3.0            SingleR_1.0.6               scater_1.14.6              </span>
<span class="co">#&gt;  [4] ggplot2_3.3.2               scran_1.14.6                SingleCellExperiment_1.8.0 </span>
<span class="co">#&gt;  [7] SummarizedExperiment_1.16.1 DelayedArray_0.12.2         BiocParallel_1.20.1        </span>
<span class="co">#&gt; [10] matrixStats_0.56.0          Biobase_2.46.0              GenomicRanges_1.38.0       </span>
<span class="co">#&gt; [13] GenomeInfoDb_1.22.0         IRanges_2.20.2              S4Vectors_0.24.3           </span>
<span class="co">#&gt; [16] BiocGenerics_0.32.0         dplyr_1.0.1                 BiocManager_1.30.10        </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] backports_1.1.8               AnnotationHub_2.18.0          BiocFileCache_1.10.2         </span>
<span class="co">#&gt;   [4] plyr_1.8.6                    igraph_1.2.5                  lazyeval_0.2.2               </span>
<span class="co">#&gt;   [7] splines_3.6.3                 GSEABase_1.48.0               shinydashboard_0.7.1         </span>
<span class="co">#&gt;  [10] listenv_0.8.0                 usethis_1.6.1                 digest_0.6.25                </span>
<span class="co">#&gt;  [13] htmltools_0.5.0               viridis_0.5.1                 fansi_0.4.1                  </span>
<span class="co">#&gt;  [16] magrittr_1.5                  memoise_1.1.0                 remotes_2.2.0                </span>
<span class="co">#&gt;  [19] limma_3.42.2                  shinyFiles_0.8.0              readr_1.3.1                  </span>
<span class="co">#&gt;  [22] globals_0.12.5                annotate_1.64.0               askpass_1.1                  </span>
<span class="co">#&gt;  [25] pkgdown_1.5.1                 prettyunits_1.1.1             colorspace_1.4-1             </span>
<span class="co">#&gt;  [28] blob_1.2.1                    rappdirs_0.3.1                xfun_0.16                    </span>
<span class="co">#&gt;  [31] callr_3.4.3                   crayon_1.3.4                  RCurl_1.98-1.2               </span>
<span class="co">#&gt;  [34] jsonlite_1.7.0                graph_1.64.0                  ape_5.4                      </span>
<span class="co">#&gt;  [37] glue_1.4.1                    gtable_0.3.0                  zlibbioc_1.32.0              </span>
<span class="co">#&gt;  [40] XVector_0.26.0                pkgbuild_1.1.0                BiocSingular_1.2.2           </span>
<span class="co">#&gt;  [43] future.apply_1.6.0            scales_1.1.1                  msigdbr_7.1.1                </span>
<span class="co">#&gt;  [46] DBI_1.1.0                     edgeR_3.28.1                  miniUI_0.1.1.1               </span>
<span class="co">#&gt;  [49] Rcpp_1.0.5                    viridisLite_0.3.0             xtable_1.8-4                 </span>
<span class="co">#&gt;  [52] progress_1.2.2                dqrng_0.2.1                   bit_4.0.3                    </span>
<span class="co">#&gt;  [55] rsvd_1.0.3                    GSVA_1.34.0                   DT_0.14.3                    </span>
<span class="co">#&gt;  [58] htmlwidgets_1.5.1             httr_1.4.2                    FNN_1.1.3                    </span>
<span class="co">#&gt;  [61] RColorBrewer_1.1-2            ellipsis_0.3.1                pkgconfig_2.0.3              </span>
<span class="co">#&gt;  [64] XML_3.99-0.3                  uwot_0.1.8                    dbplyr_1.4.4                 </span>
<span class="co">#&gt;  [67] utf8_1.1.4                    locfit_1.5-9.4                reshape2_1.4.4               </span>
<span class="co">#&gt;  [70] tidyselect_1.1.0              rlang_0.4.7                   later_1.1.0.1                </span>
<span class="co">#&gt;  [73] AnnotationDbi_1.48.0          munsell_0.5.0                 BiocVersion_3.10.1           </span>
<span class="co">#&gt;  [76] tools_3.6.3                   cli_2.0.2                     generics_0.0.2               </span>
<span class="co">#&gt;  [79] RSQLite_2.2.0                 ExperimentHub_1.12.0          devtools_2.3.1               </span>
<span class="co">#&gt;  [82] evaluate_0.14                 stringr_1.4.0                 fastmap_1.0.1                </span>
<span class="co">#&gt;  [85] yaml_2.2.1                    processx_3.4.3                knitr_1.29                   </span>
<span class="co">#&gt;  [88] bit64_4.0.2                   fs_1.5.0                      shinycssloaders_1.0.0        </span>
<span class="co">#&gt;  [91] purrr_0.3.4                   pbapply_1.4-2                 future_1.18.0                </span>
<span class="co">#&gt;  [94] nlme_3.1-148                  mime_0.9                      biomaRt_2.42.1               </span>
<span class="co">#&gt;  [97] compiler_3.6.3                shinythemes_1.1.2             beeswarm_0.2.3               </span>
<span class="co">#&gt; [100] plotly_4.9.2.1                curl_4.3                      interactiveDisplayBase_1.24.0</span>
<span class="co">#&gt; [103] testthat_2.3.2                tibble_3.0.3                  statmod_1.4.34               </span>
<span class="co">#&gt; [106] geneplotter_1.64.0            stringi_1.4.6                 highr_0.8                    </span>
<span class="co">#&gt; [109] ps_1.3.3                      RSpectra_0.16-0               desc_1.2.0                   </span>
<span class="co">#&gt; [112] lattice_0.20-41               Matrix_1.2-18                 shinyjs_1.1                  </span>
<span class="co">#&gt; [115] vctrs_0.3.2                   pillar_1.4.6                  lifecycle_0.2.0              </span>
<span class="co">#&gt; [118] BiocNeighbors_1.4.2           data.table_1.12.8             bitops_1.0-6                 </span>
<span class="co">#&gt; [121] irlba_2.3.3                   qvalue_2.18.0                 httpuv_1.5.4                 </span>
<span class="co">#&gt; [124] R6_2.4.1                      promises_1.1.1                gridExtra_2.3                </span>
<span class="co">#&gt; [127] vipor_0.4.5                   sessioninfo_1.1.1             codetools_0.2-16             </span>
<span class="co">#&gt; [130] pkgload_1.1.0                 colourpicker_1.0              MASS_7.3-51.6                </span>
<span class="co">#&gt; [133] assertthat_0.2.1              openssl_1.4.2                 rprojroot_1.3-2              </span>
<span class="co">#&gt; [136] shinyWidgets_0.5.3            withr_2.2.0                   GenomeInfoDbData_1.2.2       </span>
<span class="co">#&gt; [139] hms_0.5.3                     grid_3.6.3                    tidyr_1.1.2                  </span>
<span class="co">#&gt; [142] rmarkdown_2.3                 DelayedMatrixStats_1.8.0      shiny_1.5.0                  </span>
<span class="co">#&gt; [145] ggbeeswarm_0.7.0</span></pre></body></html></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Roman Hillje.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
